chapter04_convolutional-neural-networks/kaggle-gluon-cifar10.md

```diff
@@ -0,0 +1,412 @@
+test_dir = 'test_tiny'
+# 注意：此处对小训练集相应使用小训练批量。对Kaggle的完整训练集可设较大的整数，例如128。
+train_batch = 10
+# 注意：此处对小测试集相应使用小测试批量。对Kaggle的完整测试集可设较大的整数，例如100。
+test_batch = 1
+
+data_dir = '../data/kaggle_cifar10'
+label_file = 'trainLabels.csv'
+input_dir = 'train_valid_test_tiny'
+valid_ratio = 0.1
+reorg_cifar10_data(data_dir, label_file, train_dir, test_dir, input_dir, valid_ratio)
+```
+
+## 使用Gluon读取整理后的数据集
+
+为避免过拟合，我们在这里使用`image.CreateAugmenter`来加强数据集。例如我们设`rand_crop=True`和`rand_mirror=True`即可随机对每张图片做边切割和镜面反转。我们也通过`mean`和`std`对彩色图像RGB三个通道分别做[标准化](..chapter02_supervised-learning/kaggle-gluon-kfold.md)。以下我们列举了该函数里的所有参数，这些参数都是可以调的。
+
+```{.python .input  n=4}
+from mxnet import autograd
+from mxnet import gluon
+from mxnet import image
+from mxnet import init
+from mxnet import nd
+from mxnet.gluon.data import vision
+import numpy as np
+
+def transform(data, label):
+    im = data.astype('float32') / 255
+    auglist = image.CreateAugmenter(data_shape=(3, 32, 32), resize=0, 
+                        rand_crop=True, rand_resize=False, rand_mirror=True,
+                        mean=np.array([0.4914, 0.4822, 0.4465]), 
+                        std=np.array([0.2023, 0.1994, 0.2010]), 
+                        brightness=0, contrast=0, 
+                        saturation=0, hue=0, 
+                        pca_noise=0, rand_gray=0, inter_method=2)
+    for aug in auglist:
+        im = aug(im)
+    # 将数据格式从"高*宽*通道"改为"通道*高*宽"。
+    im = nd.transpose(im, (2,0,1))
+    return (im, nd.array([label]).asscalar().astype('float32'))
+```
+
+接下来，我们可以使用`Gluon`中的`ImageFolderDataset`类来读取整理后的数据集。
+
+```{.python .input}
+input_str = data_dir + '/' + input_dir + '/'
+
+# 读取原始图像文件。flag=1说明输入图像有三个通道（彩色）。
+train_ds = vision.ImageFolderDataset(input_str + 'train', flag=1, 
+                                     transform=transform)
+valid_ds = vision.ImageFolderDataset(input_str + 'valid', flag=1, 
+                                     transform=transform)
+train_valid_ds = vision.ImageFolderDataset(input_str + 'train_valid', 
+                                           flag=1, transform=transform)
+test_ds = vision.ImageFolderDataset(input_str + 'test', flag=1, 
+                                     transform=transform)
+
+batch_size = 100
+assert len(train_ds) >= train_batch 
+assert min(len(valid_ds), len(test_ds)) >= test_batch
+assert len(valid_ds) % test_batch == 0
+assert len(test_ds) % test_batch == 0
+
+train_data = gluon.data.DataLoader(train_ds, batch_size, shuffle=True)
+valid_data = gluon.data.DataLoader(valid_ds, batch_size, shuffle=True)
+train_valid_data = gluon.data.DataLoader(train_valid_ds, batch_size, shuffle=True)
+test_data = gluon.data.DataLoader(test_ds, batch_size, shuffle=False)
+
+# 交叉熵损失函数。
+softmax_cross_entropy = gluon.loss.SoftmaxCrossEntropyLoss()
+```
+
+## 设计模型
+
+我们这里使用了[ResNet-18](resnet-gluon.md)模型。
+
+请注意：模型可以重新设计，参数也可以重新调整。
+
+```{.python .input  n=5}
+from mxnet.gluon import nn
+from mxnet import nd
+
+class Residual(nn.Block):
```

Comment:
we may put these codes into utils.py, because i feel we are going to use it mutliple times. so we can 

`net = utils.ResNet18()`

Reply:
Glad to hear that we are going to use it multiple times. Let's provide some flexibility now.